library/std/src/io/mod.rs

```diff
@@ -587,8 +557,13 @@ pub(crate) fn default_read_buf<F>(read: F, mut cursor: BorrowedCursor<'_>) -> Re
 where
     F: FnOnce(&mut [u8]) -> Result<usize>,
 {
-    let n = read(cursor.ensure_init().init_mut())?;
-    cursor.advance(n);
+    // SAFETY: We do not uninitialize any part of the buffer.
+    let n = read(unsafe { cursor.as_mut().write_filled(0) })?;
+    assert!(n <= cursor.capacity());
+    // SAFETY: We've initialized the entire buffer, and `read` can't make it uninitialized.
+    unsafe {
+        cursor.advance(n);
+    }
```

Comment:
Too bad there is `unsafe` code here now... I really liked the fact that `read_buf` was easy to implement before.